### ISSUE TITLE
[@ngrx/data] support omitting primary key or accepting transform function

### ISSUE THREAD

**Issue by yharaskrik** [ngrx/platform#3070]
The problem:

When `PUT`ing data to the server to update it the method signature is `.update(entity: T | Partial<T>)` but the ID via `selectId` is required (since it is used in the url the http client is called with) but it should not also be in the body (since you are not actually updating it).

Currently NgRx Data just passes the entire value passed to `.update` into the HttpClient as the body. Due to some strict APIs that do not allow the primary key to be updated and cause 400 errors this does not work.

The solution:

What I propose, as to not break compatibility for anyone relying on this functionality, is to allow for one of two things.

1. add a flag to the `.update` call `omitId: boolean` that would parse out and omit the key defined by `selectId` and not pass that into the body of the request (this will require more work because right now `selectId` returns the value of the ID, not the ID itself. Not sure how best to handle that)
2. allow for a transform function to be passed in such that the body can be manipulated prior to sending off the request with the HttpClient, this can either be done at `.update` time or at entity meta data definition time. 

Neither are that preferable to be honest, open to ideas on how we can best allow for omitting an entities primary key from the update requests

## Describe any alternatives/workarounds you're currently using

None currently, other than loosening up the API to allow for having the primary key in the request and explicitly ignoring it on the server, which is not as preferable.

## Other information:

This definitely does not suit everyone's case and should not be a global change but I think it makes the most sense to allow for supporting stricter APIs or providing a mechanism for people to allow altering the request bodies prior to sending them off. This might be possible currently by overriding some service but just not sure what one, and I don't want to have to override/reimplement multiple methods just for this one use case.

## If accepted, I would be willing to submit a PR for this feature

[x] Yes (Assistance is provided if you need help submitting a pull request)
[ ] No

**Comment by markostanimirovic** [comment]
You can also extend `DefaultDataService` and override `update` method.

https://ngrx.io/guide/data/entity-dataservice#custom-entitydataservice

**Comment by yharaskrik** [comment]
@markostanimirovic that is definitely true. And I do implement custom data services in other places but I feel like (and I could be wrong here so feel free to correct me) that ngrx data advertises itself as built to follow the http standard so that everything is done the same way. It seems weird to me that the "standard" would be to allow updating the primary key of an entity, ie. Having the primary key in the PUT call. Open to being convinced otherwise though!

**Comment by donohoea** [comment]
I don't think you can actually update the primary key. As the url of the update request is determined by the primary key in the `(entity: <T> | Partial<T>)`  you can never have an update where the primary key in the data being sent to the api is different from the primary key of the api endpoint you're calling. I agree leaving the id field in the update data is redundant and I could see how a strict api might interpret that as an attempt to change the primary key even though the values are the same. I think it is this way because ngrx/data is flexible in the way it allows you to define the id field, which as you mention in your first proposed solution makes it difficult to omit the id field or fields. I agree with the @markostanimirovic the best solution would be to  extend `DefaultDataService` and override `update` method providing the transform function you proposed in solution 2. Or if it's  a common enough issue for you, extend `EntityMetadata` and `DefaultDataServiceFactory` so you can provide your transform at `EntityMetadata` definition.

**Comment by cannujBarfu0negjiz** [comment]
> I don't think you can actually update the primary key. As the url of the update request is determined by the primary key in the `(entity: <T> | Partial<T>)`  you can never have an update where the primary key in the data being sent to the api is different from the primary key of the api endpoint you're calling. I agree leaving the id field in the update data is redundant and I could see how a strict api might interpret that as an attempt to change the primary key even though the values are the same. I think it is this way because ngrx/data is flexible in the way it allows you to define the id field, which as you mention in your first proposed solution makes it difficult to omit the id field or fields. I agree with the @markostanimirovic the best solution would be to  extend `DefaultDataService` and override `update` method providing the transform function you proposed in solution 2. Or if it's  a common enough issue for you, extend `EntityMetadata` and `DefaultDataServiceFactory` so you can provide your transform at `EntityMetadata` definition.

# ****

**Comment by cannujBarfu0negjiz** [comment]
No seleccionar ningún cambio

**Comment by david-shortman** [comment]
This issue appears to be resolved by some suggested consumer implementation changes. @ngrx/data does not appear to be breaking HTTP conventions as-is.

**Comment by markostanimirovic** [comment]
I'm going to close this feature request because NgRx Data is in maintenance mode. Improvements/new features won't be added anymore.
